// aws-cdk/lib/list-stacks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listStacks = listStacks;
require("@jsii/check-node/run");
const cloud_assembly_1 = require("./api/cxapp/cloud-assembly");
/**
 * List Stacks
 *
 * @param toolkit cdk toolkit
 * @param options list stacks options
 * @returns StackDetails[]
 */
async function listStacks(toolkit, options) {
    const assembly = await toolkit.assembly();
    const stacks = await assembly.selectStacks({
        patterns: options.selectors,
    }, {
        extend: cloud_assembly_1.ExtendedStackSelection.Upstream,
        defaultBehavior: cloud_assembly_1.DefaultSelection.AllStacks,
    });
    function calculateStackDependencies(collectionOfStacks) {
        const allData = [];
        for (const stack of collectionOfStacks.stackArtifacts) {
            const data = {
                id: stack.displayName ?? stack.id,
                name: stack.stackName,
                environment: stack.environment,
                dependencies: [],
            };
            for (const dependencyId of stack.dependencies.map(x => x.id)) {
                if (dependencyId.includes('.assets')) {
                    continue;
                }
                const depStack = assembly.stackById(dependencyId);
                if (depStack.stackArtifacts[0].dependencies.length > 0 &&
                    depStack.stackArtifacts[0].dependencies.filter((dep) => !(dep.id).includes('.assets')).length > 0) {
                    const stackWithDeps = calculateStackDependencies(depStack);
                    for (const stackDetail of stackWithDeps) {
                        data.dependencies.push({
                            id: stackDetail.id,
                            dependencies: stackDetail.dependencies,
                        });
                    }
                }
                else {
                    data.dependencies.push({
                        id: depStack.stackArtifacts[0].displayName ?? depStack.stackArtifacts[0].id,
                        dependencies: [],
                    });
                }
            }
            allData.push(data);
        }
        return allData;
    }
    return calculateStackDependencies(stacks);
}
