// aws-cdk/test/api/credential-plugins.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const credential_plugins_1 = require("../../lib/api/aws-auth/credential-plugins");
const plugin_1 = require("../../lib/api/plugin");
const mode_1 = require("../../lib/api/plugin/mode");
test('returns credential from plugin', async () => {
    // GIVEN
    const creds = {
        accessKeyId: 'aaa',
        secretAccessKey: 'bbb',
        getPromise: () => Promise.resolve(),
    };
    const host = plugin_1.PluginHost.instance;
    host.registerCredentialProviderSource({
        name: 'Fake',
        canProvideCredentials(_accountId) {
            return Promise.resolve(true);
        },
        isAvailable() {
            return Promise.resolve(true);
        },
        getProvider(_accountId, _mode) {
            return Promise.resolve(creds);
        },
    });
    const plugins = new credential_plugins_1.CredentialPlugins();
    // WHEN
    const pluginCredentials = await plugins.fetchCredentialsFor('aaa', mode_1.Mode.ForReading);
    // THEN
    await expect(pluginCredentials?.credentials()).resolves.toEqual(expect.objectContaining({
        accessKeyId: 'aaa',
        secretAccessKey: 'bbb',
    }));
});
