// aws-cdk/lib/toolkit/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthenticationError = exports.ToolkitError = void 0;
const TOOLKIT_ERROR_SYMBOL = Symbol.for('@aws-cdk/core.ToolkitError');
const AUTHENTICATION_ERROR_SYMBOL = Symbol.for('@aws-cdk/core.AuthenticationError');
/**
 * Represents a general toolkit error in the AWS CDK Toolkit.
 */
class ToolkitError extends Error {
    /**
     * Determines if a given error is an instance of ToolkitError.
     */
    static isToolkitError(x) {
        return x !== null && typeof (x) === 'object' && TOOLKIT_ERROR_SYMBOL in x;
    }
    /**
     * Determines if a given error is an instance of AuthenticationError.
     */
    static isAuthenticationError(x) {
        return this.isToolkitError(x) && AUTHENTICATION_ERROR_SYMBOL in x;
    }
    constructor(message, type = 'toolkit') {
        super(message);
        Object.setPrototypeOf(this, ToolkitError.prototype);
        Object.defineProperty(this, TOOLKIT_ERROR_SYMBOL, { value: true });
        this.name = new.target.name;
        this.type = type;
    }
}
exports.ToolkitError = ToolkitError;
/**
 * Represents an authentication-specific error in the AWS CDK Toolkit.
 */
class AuthenticationError extends ToolkitError {
    constructor(message) {
        super(message, 'authentication');
        Object.setPrototypeOf(this, AuthenticationError.prototype);
        Object.defineProperty(this, AUTHENTICATION_ERROR_SYMBOL, { value: true });
    }
}
exports.AuthenticationError = AuthenticationError;
