// aws-cdk/test/bootstrap.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bootstrap_environment_1 = require("../lib/api/bootstrap/bootstrap-environment");
const cli_1 = require("../lib/cli");
beforeEach(() => {
    jest.clearAllMocks();
});
describe('cdk bootstrap', () => {
    const bootstrapEnvironmentMock = jest.spyOn(bootstrap_environment_1.Bootstrapper.prototype, 'bootstrapEnvironment');
    test('will bootstrap the a provided environment', async () => {
        bootstrapEnvironmentMock.mockResolvedValueOnce({
            noOp: false,
            outputs: {},
            type: 'did-deploy-stack',
            stackArn: 'fake-arn',
        });
        await (0, cli_1.exec)(['bootstrap', 'aws://123456789012/us-east-1']);
        expect(bootstrapEnvironmentMock).toHaveBeenCalledTimes(1);
        expect(bootstrapEnvironmentMock).toHaveBeenCalledWith({
            name: 'aws://123456789012/us-east-1',
            account: '123456789012',
            region: 'us-east-1',
        }, expect.anything(), expect.anything());
    });
});
describe('cdk bootstrap --show-template', () => {
    const stdoutSpy = jest.spyOn(process.stdout, 'write').mockImplementation(() => { return true; });
    test('prints the default bootstrap template', async () => {
        await (0, cli_1.exec)(['bootstrap', '--show-template']);
        expect(stdoutSpy).toHaveBeenCalledWith(expect.stringContaining('BootstrapVersion'));
    });
});
