// aws-cdk/test/util/mock-toolkitinfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockToolkitInfo = void 0;
/* eslint-disable import/order */
const api_1 = require("../../lib/api");
class MockToolkitInfo extends api_1.ToolkitInfo {
    static setup(toolkitInfo) {
        toolkitInfo = toolkitInfo ?? new MockToolkitInfo();
        const orig = api_1.ToolkitInfo.lookup;
        api_1.ToolkitInfo.lookup = jest.fn().mockResolvedValue(toolkitInfo);
        return {
            toolkitInfo,
            dispose: () => {
                api_1.ToolkitInfo.lookup = orig;
            },
        };
    }
    constructor(props = {}) {
        super();
        this.found = true;
        this.stackName = 'MockBootstrapStack';
        this.bucketName = props.bucketName ?? 'MockToolkitBucketName';
        this.bucketUrl = props.bucketUrl ?? `https://${this.bucketName}.s3.amazonaws.com/`;
        this.repositoryName = props.repositoryName ?? 'MockToolkitRepositoryName';
        this.version = props.version ?? 1;
        this.variant = api_1.DEFAULT_BOOTSTRAP_VARIANT;
        this._bootstrapStack = props.bootstrapStack;
    }
    get bootstrapStack() {
        if (!this._bootstrapStack) {
            throw new Error('Bootstrap stack object expected but not supplied to MockToolkitInfo');
        }
        return this._bootstrapStack;
    }
}
exports.MockToolkitInfo = MockToolkitInfo;
