// aws-cdk/test/util/stack-monitor.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const client_cloudformation_1 = require("@aws-sdk/client-cloudformation");
const mock_sdk_1 = require("./mock-sdk");
const stack_activity_monitor_1 = require("../../lib/api/util/cloudformation/stack-activity-monitor");
let sdk;
let printer;
let monitor;
beforeEach(() => {
    sdk = new mock_sdk_1.MockSdk();
    printer = new FakePrinter();
    monitor = new stack_activity_monitor_1.StackActivityMonitor(sdk.cloudFormation(), 'StackName', printer, undefined, new Date(T100)).start();
    (0, mock_sdk_1.restoreSdkMocksToDefault)();
});
describe('stack monitor event ordering and pagination', () => {
    test('continue to the next page if it exists', async () => {
        mock_sdk_1.mockCloudFormationClient.on(client_cloudformation_1.DescribeStackEventsCommand).resolvesOnce({
            StackEvents: [event(102), event(101)],
        });
        await eventually(() => expect(mock_sdk_1.mockCloudFormationClient).toHaveReceivedCommand(client_cloudformation_1.DescribeStackEventsCommand), 2);
        await monitor.stop();
        // Printer sees them in chronological order
        expect(printer.eventIds).toEqual(['101', '102']);
    });
    test('do not page further if we already saw the last event', async () => {
        mock_sdk_1.mockCloudFormationClient
            .on(client_cloudformation_1.DescribeStackEventsCommand)
            .resolvesOnce({
            StackEvents: [event(101)],
        })
            .resolvesOnce({
            StackEvents: [event(102), event(101)],
        })
            .resolvesOnce({});
        await eventually(() => expect(mock_sdk_1.mockCloudFormationClient).toHaveReceivedCommand(client_cloudformation_1.DescribeStackEventsCommand), 2);
        await monitor.stop();
        // Seen in chronological order
        expect(printer.eventIds).toEqual(['101', '102']);
    });
    test('do not page further if the last event is too old', async () => {
        mock_sdk_1.mockCloudFormationClient
            .on(client_cloudformation_1.DescribeStackEventsCommand)
            .resolvesOnce({
            StackEvents: [event(101), event(95)],
        })
            .resolvesOnce({
            StackEvents: [],
        });
        await eventually(() => expect(mock_sdk_1.mockCloudFormationClient).toHaveReceivedCommand(client_cloudformation_1.DescribeStackEventsCommand), 2);
        await monitor.stop();
        // Seen only the new one
        expect(printer.eventIds).toEqual(['101']);
    });
    test('do a final request after the monitor is stopped', async () => {
        mock_sdk_1.mockCloudFormationClient.on(client_cloudformation_1.DescribeStackEventsCommand).resolves({
            StackEvents: [event(101)],
        });
        // Establish that we've received events prior to stop and then reset the mock
        await eventually(() => expect(mock_sdk_1.mockCloudFormationClient).toHaveReceivedCommand(client_cloudformation_1.DescribeStackEventsCommand), 2);
        mock_sdk_1.mockCloudFormationClient.resetHistory();
        await monitor.stop();
        mock_sdk_1.mockCloudFormationClient.on(client_cloudformation_1.DescribeStackEventsCommand).resolves({
            StackEvents: [event(102), event(101)],
        });
        // Since we can't reset the mock to a new value before calling stop, we'll have to check
        // and make sure it's called again instead.
        expect(mock_sdk_1.mockCloudFormationClient).toHaveReceivedCommand(client_cloudformation_1.DescribeStackEventsCommand);
    });
});
describe('stack monitor, collecting errors from events', () => {
    test('return errors from the root stack', async () => {
        mock_sdk_1.mockCloudFormationClient.on(client_cloudformation_1.DescribeStackEventsCommand).resolvesOnce({
            StackEvents: [addErrorToStackEvent(event(100))],
        });
        await eventually(() => expect(mock_sdk_1.mockCloudFormationClient).toHaveReceivedCommand(client_cloudformation_1.DescribeStackEventsCommand), 2);
        await monitor.stop();
        expect(monitor.errors).toStrictEqual(['Test Error']);
    });
    test('return errors from the nested stack', async () => {
        mock_sdk_1.mockCloudFormationClient
            .on(client_cloudformation_1.DescribeStackEventsCommand)
            .resolvesOnce({
            StackEvents: [
                addErrorToStackEvent(event(102), {
                    logicalResourceId: 'nestedStackLogicalResourceId',
                    physicalResourceId: 'nestedStackPhysicalResourceId',
                    resourceType: 'AWS::CloudFormation::Stack',
                    resourceStatusReason: 'nested stack failed',
                    resourceStatus: client_cloudformation_1.ResourceStatus.UPDATE_FAILED,
                }),
                addErrorToStackEvent(event(100), {
                    logicalResourceId: 'nestedStackLogicalResourceId',
                    physicalResourceId: 'nestedStackPhysicalResourceId',
                    resourceType: 'AWS::CloudFormation::Stack',
                    resourceStatus: client_cloudformation_1.ResourceStatus.UPDATE_IN_PROGRESS,
                }),
            ],
        })
            .resolvesOnce({
            StackEvents: [
                addErrorToStackEvent(event(101), {
                    logicalResourceId: 'nestedResource',
                    resourceType: 'Some::Nested::Resource',
                    resourceStatusReason: 'actual failure error message',
                }),
            ],
        });
        await eventually(() => expect(mock_sdk_1.mockCloudFormationClient).toHaveReceivedNthCommandWith(1, client_cloudformation_1.DescribeStackEventsCommand, {
            StackName: 'StackName',
        }), 2);
        await eventually(() => expect(mock_sdk_1.mockCloudFormationClient).toHaveReceivedNthCommandWith(2, client_cloudformation_1.DescribeStackEventsCommand, {
            StackName: 'nestedStackPhysicalResourceId',
        }), 2);
        await monitor.stop();
        expect(monitor.errors).toStrictEqual(['actual failure error message', 'nested stack failed']);
    });
    test('does not consider events without physical resource id for monitoring nested stacks', async () => {
        mock_sdk_1.mockCloudFormationClient
            .on(client_cloudformation_1.DescribeStackEventsCommand)
            .resolvesOnce({
            StackEvents: [
                addErrorToStackEvent(event(100), {
                    logicalResourceId: 'nestedStackLogicalResourceId',
                    physicalResourceId: '',
                    resourceType: 'AWS::CloudFormation::Stack',
                    resourceStatusReason: 'nested stack failed',
                }),
            ],
            NextToken: 'nextToken',
        })
            .resolvesOnce({
            StackEvents: [
                addErrorToStackEvent(event(101), {
                    logicalResourceId: 'OtherResource',
                    resourceType: 'Some::Other::Resource',
                    resourceStatusReason: 'some failure',
                }),
            ],
        });
        await eventually(() => expect(mock_sdk_1.mockCloudFormationClient).toHaveReceivedCommand(client_cloudformation_1.DescribeStackEventsCommand), 2);
        await monitor.stop();
        expect(monitor.errors).toStrictEqual(['nested stack failed', 'some failure']);
        expect(mock_sdk_1.mockCloudFormationClient).toHaveReceivedNthCommandWith(1, client_cloudformation_1.DescribeStackEventsCommand, {
            StackName: 'StackName',
        });
        // Note that the second call happened for the top level stack instead of a nested stack
        expect(mock_sdk_1.mockCloudFormationClient).toHaveReceivedNthCommandWith(2, client_cloudformation_1.DescribeStackEventsCommand, {
            StackName: 'StackName',
        });
    });
    test('does not check for nested stacks that have already completed successfully', async () => {
        mock_sdk_1.mockCloudFormationClient.on(client_cloudformation_1.DescribeStackEventsCommand).resolvesOnce({
            StackEvents: [
                addErrorToStackEvent(event(100), {
                    logicalResourceId: 'nestedStackLogicalResourceId',
                    physicalResourceId: 'nestedStackPhysicalResourceId',
                    resourceType: 'AWS::CloudFormation::Stack',
                    resourceStatusReason: 'nested stack status reason',
                    resourceStatus: client_cloudformation_1.StackStatus.CREATE_COMPLETE,
                }),
            ],
        });
        await eventually(() => expect(mock_sdk_1.mockCloudFormationClient).toHaveReceivedCommand(client_cloudformation_1.DescribeStackEventsCommand), 2);
        await monitor.stop();
        expect(monitor.errors).toStrictEqual([]);
    });
});
const T0 = 1597837230504;
// Events 0-99 are before we started paying attention
const T100 = T0 + 100 * 1000;
function event(nr) {
    return {
        EventId: `${nr}`,
        StackId: 'StackId',
        StackName: 'StackName',
        Timestamp: new Date(T0 + nr * 1000),
    };
}
function addErrorToStackEvent(eventToUpdate, props = {}) {
    eventToUpdate.ResourceStatus = props.resourceStatus ?? client_cloudformation_1.ResourceStatus.UPDATE_FAILED;
    eventToUpdate.ResourceType = props.resourceType ?? 'Test::Resource::Type';
    eventToUpdate.ResourceStatusReason = props.resourceStatusReason ?? 'Test Error';
    eventToUpdate.LogicalResourceId = props.logicalResourceId ?? 'testLogicalId';
    eventToUpdate.PhysicalResourceId = props.physicalResourceId ?? 'testPhysicalResourceId';
    return eventToUpdate;
}
class FakePrinter {
    constructor() {
        this.updateSleep = 0;
        this.activities = [];
    }
    get eventIds() {
        return this.activities.map((a) => a.event.EventId);
    }
    addActivity(activity) {
        this.activities.push(activity);
    }
    print() { }
    start() { }
    stop() { }
}
const wait = () => new Promise((resolve) => setTimeout(resolve, 5));
// Using the eventually function to ensure these functions have had sufficient time to execute.
const eventually = async (call, attempts) => {
    while (attempts-- >= 0) {
        try {
            return call();
        }
        catch (err) {
            if (attempts <= 0)
                throw err;
        }
        await wait();
    }
    throw new Error('An unexpected error has occurred.');
};
