// aws-cdk/test/api/util/cloudformation/stack-event-poller.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const client_cloudformation_1 = require("@aws-sdk/client-cloudformation");
const stack_event_poller_1 = require("../../../../lib/api/util/cloudformation/stack-event-poller");
const mock_sdk_1 = require("../../../util/mock-sdk");
beforeEach(() => {
    jest.resetAllMocks();
});
describe('poll', () => {
    test('polls all necessary pages', async () => {
        const deployTime = Date.now();
        const postDeployEvent1 = {
            Timestamp: new Date(deployTime + 1000),
            EventId: 'event-1',
            StackId: 'stack-id',
            StackName: 'stack',
        };
        const postDeployEvent2 = {
            Timestamp: new Date(deployTime + 2000),
            EventId: 'event-2',
            StackId: 'stack-id',
            StackName: 'stack',
        };
        const sdk = new mock_sdk_1.MockSdk();
        mock_sdk_1.mockCloudFormationClient.on(client_cloudformation_1.DescribeStackEventsCommand).callsFake((input) => {
            const result = {
                StackEvents: input.NextToken === 'token' ? [postDeployEvent2] : [postDeployEvent1],
                NextToken: input.NextToken === 'token' ? undefined : 'token', // simulate a two page event stream.
            };
            return result;
        });
        const poller = new stack_event_poller_1.StackEventPoller(sdk.cloudFormation(), {
            stackName: 'stack',
            startTime: new Date().getTime(),
        });
        const events = await poller.poll();
        expect(events.length).toEqual(2);
    });
    test('does not poll unnecessary pages', async () => {
        const deployTime = Date.now();
        const preDeployTimeEvent = {
            Timestamp: new Date(deployTime - 1000),
            EventId: 'event-1',
            StackId: 'stack-id',
            StackName: 'stack',
        };
        const sdk = new mock_sdk_1.MockSdk();
        mock_sdk_1.mockCloudFormationClient.on(client_cloudformation_1.DescribeStackEventsCommand).callsFake((input) => {
            // the first event we return should stop the polling. we therefore
            // do not expect a second page to be polled.
            expect(input.NextToken).toBe(undefined);
            return {
                StackEvents: [preDeployTimeEvent],
                NextToken: input.NextToken === 'token' ? undefined : 'token', // simulate a two page event stream.
            };
        });
        const poller = new stack_event_poller_1.StackEventPoller(sdk.cloudFormation(), {
            stackName: 'stack',
            startTime: new Date().getTime(),
        });
        await poller.poll();
    });
});
