// aws-cdk/test/api/aws-auth/sdk-logger.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_logger_1 = require("../../../lib/api/aws-auth/sdk-logger");
const logging = require("../../../lib/logging");
describe(sdk_logger_1.SdkToCliLogger, () => {
    const logger = new sdk_logger_1.SdkToCliLogger();
    const trace = jest.spyOn(logging, 'trace');
    beforeEach(() => {
        trace.mockReset();
    });
    test.each(['trace', 'debug'])('%s method does not call trace', (meth) => {
        logger[meth]('test');
        expect(trace).not.toHaveBeenCalled();
    });
    test.each(['info', 'warn', 'error'])('%s method logs to trace', (meth) => {
        logger[meth]('test');
        expect(trace).toHaveBeenCalled();
    });
});
const SUCCESS_CALL = {
    clientName: 'S3Client',
    commandName: 'GetBucketLocationCommand',
    input: {
        Bucket: '.....',
        ExpectedBucketOwner: undefined,
    },
    output: { LocationConstraint: 'eu-central-1' },
    metadata: {
        httpStatusCode: 200,
        requestId: '....',
        extendedRequestId: '...',
        cfId: undefined,
        attempts: 2,
        totalRetryDelay: 30,
    },
};
const ERROR_CALL = {
    clientName: 'S3Client',
    commandName: 'GetBucketLocationCommand',
    input: {
        Bucket: '.....',
        ExpectedBucketOwner: undefined,
    },
    error: new Error('it failed'),
    metadata: {
        httpStatusCode: 200,
        attempts: 2,
        totalRetryDelay: 30,
    },
};
test('formatting a successful SDK call looks broadly reasonable', () => {
    const output = (0, sdk_logger_1.formatSdkLoggerContent)([SUCCESS_CALL]);
    expect(output).toMatchSnapshot();
});
test('formatting a failing SDK call looks broadly reasonable', () => {
    const output = (0, sdk_logger_1.formatSdkLoggerContent)([ERROR_CALL]);
    expect(output).toMatchSnapshot();
});
test('formatting a successful SDK call includes attempts and retries if greater than 1', () => {
    const output = (0, sdk_logger_1.formatSdkLoggerContent)([SUCCESS_CALL]);
    expect(output).toContain('2 attempts');
    expect(output).toContain('30ms');
});
test('formatting a failing SDK call includes attempts and retries if greater than 1', () => {
    const output = (0, sdk_logger_1.formatSdkLoggerContent)([ERROR_CALL]);
    expect(output).toContain('2 attempts');
    expect(output).toContain('30ms');
});
test('formatting a failing SDK call includes the error', () => {
    const output = (0, sdk_logger_1.formatSdkLoggerContent)([ERROR_CALL]);
    expect(output).toContain('it failed');
});
