// aws-cdk/test/context-providers/amis.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("aws-sdk-client-mock");
const client_ec2_1 = require("@aws-sdk/client-ec2");
const lib_1 = require("../../lib");
const ami_1 = require("../../lib/context-providers/ami");
const mock_sdk_1 = require("../util/mock-sdk");
const mockSDK = new (class extends mock_sdk_1.MockSdkProvider {
    forEnvironment() {
        return Promise.resolve({ sdk: new lib_1.SDK(mock_sdk_1.FAKE_CREDENTIAL_CHAIN, mockSDK.defaultRegion, {}), didAssumeRole: false });
    }
})();
test('calls DescribeImages on the request', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client.on(client_ec2_1.DescribeImagesCommand).resolves({
        Images: [{ ImageId: 'ami-1234' }],
    });
    // WHEN
    await new ami_1.AmiContextProviderPlugin(mockSDK).getValue({
        account: '1234',
        region: 'asdf',
        owners: ['some-owner'],
        filters: {
            'some-filter': ['filtered'],
        },
    });
    // THEN
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeImagesCommand, {
        Owners: ['some-owner'],
        Filters: [
            {
                Name: 'some-filter',
                Values: ['filtered'],
            },
        ],
    });
});
test('returns the most recent AMI matching the criteria', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client.on(client_ec2_1.DescribeImagesCommand).resolves({
        Images: [
            {
                ImageId: 'ami-1234',
                CreationDate: '2016-06-22T08:39:59.000Z',
            },
            {
                ImageId: 'ami-5678',
                CreationDate: '2019-06-22T08:39:59.000Z',
            },
        ],
    });
    // WHEN
    const result = await new ami_1.AmiContextProviderPlugin(mockSDK).getValue({
        account: '1234',
        region: 'asdf',
        filters: {},
    });
    // THEN
    expect(result).toBe('ami-5678');
});
