// aws-cdk/test/context-providers/endpoint-service-availability-zones.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const client_ec2_1 = require("@aws-sdk/client-ec2");
const lib_1 = require("../../lib");
const endpoint_service_availability_zones_1 = require("../../lib/context-providers/endpoint-service-availability-zones");
const mock_sdk_1 = require("../util/mock-sdk");
const mockSDK = new (class extends mock_sdk_1.MockSdkProvider {
    forEnvironment() {
        return Promise.resolve({ sdk: new lib_1.SDK(mock_sdk_1.FAKE_CREDENTIAL_CHAIN, mockSDK.defaultRegion, {}), didAssumeRole: false });
    }
})();
test('empty result when service details cannot be retrieved', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client.on(client_ec2_1.DescribeVpcEndpointServicesCommand).resolves({});
    // WHEN
    const result = await new endpoint_service_availability_zones_1.EndpointServiceAZContextProviderPlugin(mockSDK).getValue({
        serviceName: 'svc',
        account: 'foo',
        region: 'rgn',
    });
    expect(result).toEqual([]);
});
test('returns availability zones', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client.on(client_ec2_1.DescribeVpcEndpointServicesCommand).resolves({
        ServiceDetails: [{
                AvailabilityZones: ['us-east-1a'],
            }],
    });
    // WHEN
    const result = await new endpoint_service_availability_zones_1.EndpointServiceAZContextProviderPlugin(mockSDK).getValue({
        serviceName: 'svc',
        account: 'foo',
        region: 'rgn',
    });
    expect(result).toEqual(['us-east-1a']);
});
