// aws-cdk/test/context-providers/hosted-zones.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const client_route_53_1 = require("@aws-sdk/client-route-53");
const lib_1 = require("../../lib");
const hosted_zones_1 = require("../../lib/context-providers/hosted-zones");
const mock_sdk_1 = require("../util/mock-sdk");
const mockSDK = new (class extends mock_sdk_1.MockSdkProvider {
    forEnvironment() {
        return Promise.resolve({ sdk: new lib_1.SDK(mock_sdk_1.FAKE_CREDENTIAL_CHAIN, mockSDK.defaultRegion, {}), didAssumeRole: false });
    }
})();
test('get value without private zone', async () => {
    // GIVEN
    mock_sdk_1.mockRoute53Client.on(client_route_53_1.ListHostedZonesByNameCommand).resolves({
        HostedZones: [{
                Id: 'foo',
                Name: 'example.com.',
                CallerReference: 'xyz',
            }],
    });
    // WHEN
    const result = await new hosted_zones_1.HostedZoneContextProviderPlugin(mockSDK).getValue({
        domainName: 'example.com',
        account: '1234',
        region: 'rgn',
    });
    expect(result).toEqual({
        Id: 'foo',
        Name: 'example.com.',
    });
});
test('get value with private zone', async () => {
    // GIVEN
    mock_sdk_1.mockRoute53Client.on(client_route_53_1.ListHostedZonesByNameCommand).resolves({
        HostedZones: [{
                Id: 'foo',
                Name: 'example.com.',
                CallerReference: 'xyz',
                Config: {
                    PrivateZone: true,
                },
            }],
    });
    // WHEN
    const result = await new hosted_zones_1.HostedZoneContextProviderPlugin(mockSDK).getValue({
        domainName: 'example.com',
        account: '1234',
        region: 'rgn',
        privateZone: true,
    });
    expect(result).toEqual({
        Id: 'foo',
        Name: 'example.com.',
    });
});
test('get value with private zone and VPC not found', async () => {
    // GIVEN
    mock_sdk_1.mockRoute53Client.on(client_route_53_1.ListHostedZonesByNameCommand).resolves({
        HostedZones: [{
                Id: 'foo',
                Name: 'example.com.',
                CallerReference: 'xyz',
                Config: {
                    PrivateZone: true,
                },
            }],
    });
    // No VPCs
    mock_sdk_1.mockRoute53Client.on(client_route_53_1.GetHostedZoneCommand).resolves({});
    // WHEN
    const result = new hosted_zones_1.HostedZoneContextProviderPlugin(mockSDK).getValue({
        domainName: 'example.com',
        account: '1234',
        region: 'rgn',
        privateZone: true,
        vpcId: 'vpc-bla',
    });
    await expect(result)
        .rejects
        .toThrow(new Error('Found zones: [] for dns:example.com, privateZone:true, vpcId:vpc-bla, but wanted exactly 1 zone'));
});
test('get value with private zone and VPC found', async () => {
    // GIVEN
    mock_sdk_1.mockRoute53Client.on(client_route_53_1.ListHostedZonesByNameCommand).resolves({
        HostedZones: [{
                Id: 'foo',
                Name: 'example.com.',
                CallerReference: 'xyz',
                Config: {
                    PrivateZone: true,
                },
            }],
    });
    mock_sdk_1.mockRoute53Client.on(client_route_53_1.GetHostedZoneCommand).resolves({
        VPCs: [{
                VPCId: 'vpc-bla',
            }],
    });
    // WHEN
    const result = await new hosted_zones_1.HostedZoneContextProviderPlugin(mockSDK).getValue({
        domainName: 'example.com',
        account: '1234',
        region: 'rgn',
        privateZone: true,
        vpcId: 'vpc-bla',
    });
    expect(result).toEqual({
        Id: 'foo',
        Name: 'example.com.',
    });
});
