// aws-cdk/test/context-providers/load-balancers.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cloud_assembly_schema_1 = require("@aws-cdk/cloud-assembly-schema");
const client_elastic_load_balancing_v2_1 = require("@aws-sdk/client-elastic-load-balancing-v2");
const lib_1 = require("../../lib");
const load_balancers_1 = require("../../lib/context-providers/load-balancers");
const mock_sdk_1 = require("../util/mock-sdk");
const mockSDK = new (class extends mock_sdk_1.MockSdkProvider {
    forEnvironment() {
        return Promise.resolve({ sdk: new lib_1.SDK(mock_sdk_1.FAKE_CREDENTIAL_CHAIN, mockSDK.defaultRegion, {}), didAssumeRole: false });
    }
})();
beforeEach(() => {
    (0, mock_sdk_1.restoreSdkMocksToDefault)();
});
describe('load balancer context provider plugin', () => {
    test('errors when no matches are found', async () => {
        // GIVEN
        const provider = new load_balancers_1.LoadBalancerContextProviderPlugin(mockSDK);
        // WHEN
        await expect(provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
            loadBalancerArn: 'arn:load-balancer1',
        })).rejects.toThrow('No load balancers found matching {"account":"1234","region":"us-east-1","loadBalancerType":"application","loadBalancerArn":"arn:load-balancer1"}');
    });
    test('errors when multiple load balancers match', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client
            .on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand)
            .resolves({
            LoadBalancers: [
                {
                    IpAddressType: 'ipv4',
                    LoadBalancerArn: 'arn:load-balancer1',
                    DNSName: 'dns1.example.com',
                    CanonicalHostedZoneId: 'Z1234',
                    SecurityGroups: ['sg-1234'],
                    VpcId: 'vpc-1234',
                    Type: 'application',
                },
                {
                    IpAddressType: 'ipv4',
                    LoadBalancerArn: 'arn:load-balancer2',
                    DNSName: 'dns2.example.com',
                    CanonicalHostedZoneId: 'Z1234',
                    SecurityGroups: ['sg-1234'],
                    VpcId: 'vpc-1234',
                    Type: 'application',
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeTagsCommand)
            .resolves({
            TagDescriptions: [
                {
                    ResourceArn: 'arn:load-balancer1',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
                {
                    ResourceArn: 'arn:load-balancer2',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerContextProviderPlugin(mockSDK);
        // WHEN
        await expect(provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
            loadBalancerTags: [{ key: 'some', value: 'tag' }],
        })).rejects.toThrow('Multiple load balancers found matching {"account":"1234","region":"us-east-1","loadBalancerType":"application","loadBalancerTags":[{"key":"some","value":"tag"}]} - please provide more specific criteria');
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeTagsCommand, {
            ResourceArns: ['arn:load-balancer1', 'arn:load-balancer2'],
        });
    });
    test('looks up by arn', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client.on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand).resolves({
            LoadBalancers: [
                {
                    IpAddressType: 'ipv4',
                    LoadBalancerArn: 'arn:load-balancer1',
                    DNSName: 'dns.example.com',
                    CanonicalHostedZoneId: 'Z1234',
                    SecurityGroups: ['sg-1234'],
                    VpcId: 'vpc-1234',
                    Type: 'application',
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerContextProviderPlugin(mockSDK);
        // WHEN
        const result = await provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
            loadBalancerArn: 'arn:load-balancer1',
        });
        // THEN
        expect(result.ipAddressType).toEqual('ipv4');
        expect(result.loadBalancerArn).toEqual('arn:load-balancer1');
        expect(result.loadBalancerCanonicalHostedZoneId).toEqual('Z1234');
        expect(result.loadBalancerDnsName).toEqual('dns.example.com');
        expect(result.securityGroupIds).toEqual(['sg-1234']);
        expect(result.vpcId).toEqual('vpc-1234');
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand, {
            LoadBalancerArns: ['arn:load-balancer1'],
        });
    });
    test('looks up by tags', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client
            .on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand)
            .resolves({
            LoadBalancers: [
                {
                    IpAddressType: 'ipv4',
                    LoadBalancerArn: 'arn:load-balancer1',
                    DNSName: 'dns1.example.com',
                    CanonicalHostedZoneId: 'Z1234',
                    SecurityGroups: ['sg-1234'],
                    VpcId: 'vpc-1234',
                    Type: 'application',
                },
                {
                    IpAddressType: 'ipv4',
                    LoadBalancerArn: 'arn:load-balancer2',
                    DNSName: 'dns2.example.com',
                    CanonicalHostedZoneId: 'Z1234',
                    SecurityGroups: ['sg-1234'],
                    VpcId: 'vpc-1234',
                    Type: 'application',
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeTagsCommand)
            .resolves({
            TagDescriptions: [
                {
                    ResourceArn: 'arn:load-balancer1',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
                {
                    ResourceArn: 'arn:load-balancer2',
                    Tags: [
                        { Key: 'some', Value: 'tag' },
                        { Key: 'second', Value: 'tag2' },
                    ],
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerContextProviderPlugin(mockSDK);
        // WHEN
        const result = await provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
            loadBalancerTags: [
                { key: 'some', value: 'tag' },
                { key: 'second', value: 'tag2' },
            ],
        });
        expect(result.loadBalancerArn).toEqual('arn:load-balancer2');
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeTagsCommand, {
            ResourceArns: ['arn:load-balancer1', 'arn:load-balancer2'],
        });
    });
    test('looks up by tags - query by subset', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client
            .on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand)
            .resolves({
            LoadBalancers: [
                {
                    IpAddressType: 'ipv4',
                    LoadBalancerArn: 'arn:load-balancer2',
                    DNSName: 'dns2.example.com',
                    CanonicalHostedZoneId: 'Z1234',
                    SecurityGroups: ['sg-1234'],
                    VpcId: 'vpc-1234',
                    Type: 'application',
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeTagsCommand)
            .resolves({
            TagDescriptions: [
                {
                    ResourceArn: 'arn:load-balancer2',
                    Tags: [
                        // Load balancer has two tags...
                        { Key: 'some', Value: 'tag' },
                        { Key: 'second', Value: 'tag2' },
                    ],
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerContextProviderPlugin(mockSDK);
        // WHEN
        const result = await provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
            loadBalancerTags: [
                // ...but we are querying for only one of them
                { key: 'second', value: 'tag2' },
            ],
        });
        expect(result.loadBalancerArn).toEqual('arn:load-balancer2');
    });
    test('filters by type', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client
            .on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand)
            .resolves({
            LoadBalancers: [
                {
                    IpAddressType: 'ipv4',
                    Type: 'network',
                    LoadBalancerArn: 'arn:load-balancer1',
                    DNSName: 'dns1.example.com',
                    CanonicalHostedZoneId: 'Z1234',
                    SecurityGroups: ['sg-1234'],
                    VpcId: 'vpc-1234',
                },
                {
                    IpAddressType: 'ipv4',
                    Type: 'application',
                    LoadBalancerArn: 'arn:load-balancer2',
                    DNSName: 'dns2.example.com',
                    CanonicalHostedZoneId: 'Z1234',
                    SecurityGroups: ['sg-1234'],
                    VpcId: 'vpc-1234',
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeTagsCommand)
            .resolves({
            TagDescriptions: [
                {
                    ResourceArn: 'arn:load-balancer1',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
                {
                    ResourceArn: 'arn:load-balancer2',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerContextProviderPlugin(mockSDK);
        // WHEN
        const loadBalancer = await provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerTags: [{ key: 'some', value: 'tag' }],
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
        });
        expect(loadBalancer.loadBalancerArn).toEqual('arn:load-balancer2');
    });
});
describe('load balancer listener context provider plugin', () => {
    test('errors when no associated load balancers match', async () => {
        // GIVEN
        const provider = new load_balancers_1.LoadBalancerListenerContextProviderPlugin(mockSDK);
        // WHEN
        await expect(provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
            loadBalancerTags: [{ key: 'some', value: 'tag' }],
        })).rejects.toThrow('No associated load balancers found for load balancer listener query {"account":"1234","region":"us-east-1","loadBalancerType":"application","loadBalancerTags":[{"key":"some","value":"tag"}]}');
    });
    test('errors when no listeners match', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client
            .on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand)
            .resolves({
            LoadBalancers: [
                {
                    LoadBalancerArn: 'arn:load-balancer',
                    Type: 'application',
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeListenersCommand)
            .resolves({
            Listeners: [
                {
                    LoadBalancerArn: 'arn:load-balancer',
                    ListenerArn: 'arn:listener',
                    Port: 80,
                    Protocol: 'HTTP',
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerListenerContextProviderPlugin(mockSDK);
        // WHEN
        await expect(provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
            loadBalancerArn: 'arn:load-balancer',
            listenerPort: 443,
            listenerProtocol: cloud_assembly_schema_1.LoadBalancerListenerProtocol.HTTPS,
        })).rejects.toThrow('No load balancer listeners found matching {"account":"1234","region":"us-east-1","loadBalancerType":"application","loadBalancerArn":"arn:load-balancer","listenerPort":443,"listenerProtocol":"HTTPS"}');
    });
    test('errors when multiple listeners match', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client
            .on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand)
            .resolves({
            LoadBalancers: [
                {
                    LoadBalancerArn: 'arn:load-balancer',
                    Type: 'application',
                },
                {
                    LoadBalancerArn: 'arn:load-balancer2',
                    Type: 'application',
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeTagsCommand)
            .resolves({
            TagDescriptions: [
                {
                    ResourceArn: 'arn:load-balancer',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
                {
                    ResourceArn: 'arn:load-balancer2',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeListenersCommand)
            .resolves({
            Listeners: [
                {
                    LoadBalancerArn: 'arn:load-balancer',
                    ListenerArn: 'arn:listener',
                    Port: 80,
                    Protocol: 'HTTP',
                },
                {
                    LoadBalancerArn: 'arn:load-balancer2',
                    ListenerArn: 'arn:listener2',
                    Port: 80,
                    Protocol: 'HTTP',
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerListenerContextProviderPlugin(mockSDK);
        // WHEN
        await expect(provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
            loadBalancerTags: [{ key: 'some', value: 'tag' }],
            listenerPort: 80,
            listenerProtocol: cloud_assembly_schema_1.LoadBalancerListenerProtocol.HTTP,
        })).rejects.toThrow('Multiple load balancer listeners found matching {"account":"1234","region":"us-east-1","loadBalancerType":"application","loadBalancerTags":[{"key":"some","value":"tag"}],"listenerPort":80,"listenerProtocol":"HTTP"} - please provide more specific criteria');
    });
    test('looks up by listener arn', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client
            .on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand)
            .resolves({
            LoadBalancers: [
                {
                    LoadBalancerArn: 'arn:load-balancer-arn',
                    SecurityGroups: ['sg-1234', 'sg-2345'],
                    Type: 'application',
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeListenersCommand)
            .resolves({
            Listeners: [
                {
                    ListenerArn: 'arn:listener-arn',
                    LoadBalancerArn: 'arn:load-balancer-arn',
                    Port: 999,
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerListenerContextProviderPlugin(mockSDK);
        // WHEN
        const listener = await provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
            listenerArn: 'arn:listener-arn',
        });
        // THEN
        expect(listener.listenerArn).toEqual('arn:listener-arn');
        expect(listener.listenerPort).toEqual(999);
        expect(listener.securityGroupIds).toEqual(['sg-1234', 'sg-2345']);
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand, {
            LoadBalancerArns: ['arn:load-balancer-arn'],
        });
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeListenersCommand, {
            ListenerArns: ['arn:listener-arn'],
        });
    });
    test('looks up by associated load balancer arn', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client
            .on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand)
            .resolves({
            LoadBalancers: [
                {
                    LoadBalancerArn: 'arn:load-balancer-arn1',
                    SecurityGroups: ['sg-1234'],
                    Type: 'application',
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeListenersCommand)
            .resolves({
            Listeners: [
                {
                    // This one
                    ListenerArn: 'arn:listener-arn1',
                    LoadBalancerArn: 'arn:load-balancer-arn1',
                    Port: 80,
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerListenerContextProviderPlugin(mockSDK);
        // WHEN
        const listener = await provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
            loadBalancerArn: 'arn:load-balancer-arn1',
        });
        // THEN
        expect(listener.listenerArn).toEqual('arn:listener-arn1');
        expect(listener.listenerPort).toEqual(80);
        expect(listener.securityGroupIds).toEqual(['sg-1234']);
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand, {
            LoadBalancerArns: ['arn:load-balancer-arn1'],
        });
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeListenersCommand, {
            LoadBalancerArn: 'arn:load-balancer-arn1',
        });
    });
    test('looks up by associated load balancer tags', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client
            .on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand)
            .resolves({
            LoadBalancers: [
                {
                    // This one should have the wrong tags
                    LoadBalancerArn: 'arn:load-balancer-arn1',
                    SecurityGroups: ['sg-1234', 'sg-2345'],
                    Type: 'application',
                },
                {
                    // Expecting this one
                    LoadBalancerArn: 'arn:load-balancer-arn2',
                    SecurityGroups: ['sg-3456', 'sg-4567'],
                    Type: 'application',
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeListenersCommand)
            .resolves({
            Listeners: [
                {
                    // This one
                    ListenerArn: 'arn:listener-arn1',
                    LoadBalancerArn: 'arn:load-balancer-arn2',
                    Port: 80,
                },
                {
                    ListenerArn: 'arn:listener-arn2',
                    LoadBalancerArn: 'arn:load-balancer-arn2',
                    Port: 999,
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeTagsCommand)
            .resolves({
            TagDescriptions: [
                {
                    ResourceArn: 'arn:load-balancer-arn1',
                    Tags: [],
                },
                {
                    // Expecting this one
                    ResourceArn: 'arn:load-balancer-arn2',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerListenerContextProviderPlugin(mockSDK);
        // WHEN
        const listener = await provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
            loadBalancerTags: [{ key: 'some', value: 'tag' }],
            listenerPort: 999,
        });
        // THEN
        expect(listener.listenerArn).toEqual('arn:listener-arn2');
        expect(listener.listenerPort).toEqual(999);
        expect(listener.securityGroupIds).toEqual(['sg-3456', 'sg-4567']);
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand, {
            LoadBalancerArns: undefined,
        });
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeListenersCommand, {
            LoadBalancerArn: 'arn:load-balancer-arn2',
        });
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeTagsCommand, {
            ResourceArns: ['arn:load-balancer-arn1', 'arn:load-balancer-arn2'],
        });
    });
    test('looks up by listener port and protocol', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client
            .on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand)
            .resolves({
            LoadBalancers: [
                {
                    // Shouldn't have any matching listeners
                    IpAddressType: 'ipv4',
                    LoadBalancerArn: 'arn:load-balancer1',
                    DNSName: 'dns1.example.com',
                    CanonicalHostedZoneId: 'Z1234',
                    SecurityGroups: ['sg-1234'],
                    VpcId: 'vpc-1234',
                    Type: 'application',
                },
                {
                    // Should have a matching listener
                    IpAddressType: 'ipv4',
                    LoadBalancerArn: 'arn:load-balancer2',
                    DNSName: 'dns2.example.com',
                    CanonicalHostedZoneId: 'Z1234',
                    SecurityGroups: ['sg-2345'],
                    VpcId: 'vpc-1234',
                    Type: 'application',
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeTagsCommand)
            .resolves({
            TagDescriptions: [
                {
                    ResourceArn: 'arn:load-balancer1',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
                {
                    ResourceArn: 'arn:load-balancer2',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeListenersCommand)
            .resolvesOnce({
            Listeners: [
                {
                    // Wrong port, wrong protocol => no match
                    ListenerArn: 'arn:listener-arn1',
                    LoadBalancerArn: 'arn:load-balancer1',
                    Protocol: 'HTTP',
                    Port: 80,
                },
                {
                    // Wrong protocol, right port => no match
                    ListenerArn: 'arn:listener-arn3',
                    LoadBalancerArn: 'arn:load-balancer1',
                    Protocol: 'HTTPS',
                    Port: 443,
                },
                {
                    // Wrong port, right protocol => no match
                    ListenerArn: 'arn:listener-arn4',
                    LoadBalancerArn: 'arn:load-balancer1',
                    Protocol: 'TCP',
                    Port: 999,
                },
            ],
        })
            .resolvesOnce({
            Listeners: [
                {
                    // Wrong port, wrong protocol => no match
                    ListenerArn: 'arn:listener-arn5',
                    LoadBalancerArn: 'arn:load-balancer2',
                    Protocol: 'HTTP',
                    Port: 80,
                },
                {
                    // Right port, right protocol => match
                    ListenerArn: 'arn:listener-arn6',
                    LoadBalancerArn: 'arn:load-balancer2',
                    Port: 443,
                    Protocol: 'TCP',
                },
            ],
        });
        mock_sdk_1.mockElasticLoadBalancingV2Client.on(client_elastic_load_balancing_v2_1.DescribeTagsCommand).resolves({
            TagDescriptions: [
                {
                    ResourceArn: 'arn:load-balancer1',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
                {
                    ResourceArn: 'arn:load-balancer2',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerListenerContextProviderPlugin(mockSDK);
        // WHEN
        const listener = await provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.APPLICATION,
            loadBalancerTags: [{ key: 'some', value: 'tag' }],
            listenerProtocol: cloud_assembly_schema_1.LoadBalancerListenerProtocol.TCP,
            listenerPort: 443,
        });
        // THEN
        expect(listener.listenerArn).toEqual('arn:listener-arn6');
        expect(listener.listenerPort).toEqual(443);
        expect(listener.securityGroupIds).toEqual(['sg-2345']);
    });
    test('filters by associated load balancer type', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client
            .on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand)
            .resolves({
            LoadBalancers: [
                {
                    // This one has wrong type => no match
                    LoadBalancerArn: 'arn:load-balancer-arn1',
                    SecurityGroups: [],
                    Type: 'application',
                },
                {
                    // Right type => match
                    LoadBalancerArn: 'arn:load-balancer-arn2',
                    SecurityGroups: [],
                    Type: 'network',
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeListenersCommand)
            .resolves({
            Listeners: [
                {
                    ListenerArn: 'arn:listener-arn2',
                    LoadBalancerArn: 'arn:load-balancer-arn2',
                    Port: 443,
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeTagsCommand)
            .resolves({
            TagDescriptions: [
                {
                    ResourceArn: 'arn:load-balancer-arn1',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
                {
                    ResourceArn: 'arn:load-balancer-arn2',
                    Tags: [{ Key: 'some', Value: 'tag' }],
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerListenerContextProviderPlugin(mockSDK);
        // WHEN
        const listener = await provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.NETWORK,
            loadBalancerTags: [{ key: 'some', value: 'tag' }],
            listenerPort: 443,
        });
        // THEN
        expect(listener.listenerArn).toEqual('arn:listener-arn2');
        expect(listener.listenerPort).toEqual(443);
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand, {
            LoadBalancerArns: undefined,
        });
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeListenersCommand, {
            LoadBalancerArn: 'arn:load-balancer-arn2',
        });
    });
    test('errors when associated load balancer is wrong type', async () => {
        // GIVEN
        mock_sdk_1.mockElasticLoadBalancingV2Client
            .on(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand)
            .resolves({
            LoadBalancers: [
                {
                    // This one has wrong type => no match
                    LoadBalancerArn: 'arn:load-balancer-arn1',
                    SecurityGroups: [],
                    Type: 'application',
                },
            ],
        })
            .on(client_elastic_load_balancing_v2_1.DescribeListenersCommand)
            .resolves({
            Listeners: [
                {
                    ListenerArn: 'arn:listener-arn1',
                    LoadBalancerArn: 'arn:load-balancer-arn1',
                    Port: 443,
                },
            ],
        });
        const provider = new load_balancers_1.LoadBalancerListenerContextProviderPlugin(mockSDK);
        // WHEN
        await expect(provider.getValue({
            account: '1234',
            region: 'us-east-1',
            loadBalancerType: cloud_assembly_schema_1.LoadBalancerType.NETWORK,
            listenerArn: 'arn:listener-arn1',
        })).rejects.toThrow('No associated load balancer found for listener arn arn:listener-arn1');
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeLoadBalancersCommand, {
            LoadBalancerArns: ['arn:load-balancer-arn1'],
        });
        expect(mock_sdk_1.mockElasticLoadBalancingV2Client).toHaveReceivedCommandWith(client_elastic_load_balancing_v2_1.DescribeListenersCommand, {
            ListenerArns: ['arn:listener-arn1'],
        });
    });
});
