// aws-cdk/test/context-providers/ssm-parameters.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const client_ssm_1 = require("@aws-sdk/client-ssm");
const lib_1 = require("../../lib");
const ssm_parameters_1 = require("../../lib/context-providers/ssm-parameters");
const mock_sdk_1 = require("../util/mock-sdk");
const mockSDK = new (class extends mock_sdk_1.MockSdkProvider {
    forEnvironment() {
        return Promise.resolve({ sdk: new lib_1.SDK(mock_sdk_1.FAKE_CREDENTIAL_CHAIN, mockSDK.defaultRegion, {}), didAssumeRole: false });
    }
})();
describe('ssmParameters', () => {
    test('returns value', async () => {
        (0, mock_sdk_1.restoreSdkMocksToDefault)();
        const provider = new ssm_parameters_1.SSMContextProviderPlugin(mockSDK);
        mock_sdk_1.mockSSMClient.on(client_ssm_1.GetParameterCommand).resolves({
            Parameter: {
                Value: 'bar',
            },
        });
        // WHEN
        const value = await provider.getValue({
            account: '1234',
            region: 'us-east-1',
            parameterName: 'foo',
        });
        expect(value).toEqual('bar');
    });
    test('errors when parameter is not found', async () => {
        (0, mock_sdk_1.restoreSdkMocksToDefault)();
        const provider = new ssm_parameters_1.SSMContextProviderPlugin(mockSDK);
        const notFound = new Error('Parameter not found');
        notFound.name = 'ParameterNotFound';
        mock_sdk_1.mockSSMClient.on(client_ssm_1.GetParameterCommand).rejects(notFound);
        // WHEN
        await expect(provider.getValue({
            account: '1234',
            region: 'us-east-1',
            parameterName: 'foo',
        })).rejects.toThrow(/SSM parameter not available in account/);
    });
});
