// aws-cdk/test/context-providers/vpcs.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const client_ec2_1 = require("@aws-sdk/client-ec2");
const vpcs_1 = require("../../lib/context-providers/vpcs");
const mock_sdk_1 = require("../util/mock-sdk");
const mockSDK = new mock_sdk_1.MockSdkProvider();
beforeEach(() => {
    (0, mock_sdk_1.restoreSdkMocksToDefault)();
    mock_sdk_1.mockEC2Client
        .on(client_ec2_1.DescribeVpcsCommand)
        .resolves({
        Vpcs: [{ VpcId: 'vpc-1234567', CidrBlock: '1.1.1.1/16', OwnerId: '123456789012' }],
    })
        .on(client_ec2_1.DescribeSubnetsCommand)
        .resolves({
        Subnets: [
            { SubnetId: 'sub-123456', AvailabilityZone: 'bermuda-triangle-1337', MapPublicIpOnLaunch: true },
            { SubnetId: 'sub-789012', AvailabilityZone: 'bermuda-triangle-1337', MapPublicIpOnLaunch: false },
        ],
    })
        .on(client_ec2_1.DescribeRouteTablesCommand)
        .resolves({
        RouteTables: [
            {
                Associations: [{ SubnetId: 'sub-123456' }],
                RouteTableId: 'rtb-123456',
                Routes: [
                    {
                        DestinationCidrBlock: '1.1.1.1/24',
                        GatewayId: 'local',
                        Origin: 'CreateRouteTable',
                        State: 'active',
                    },
                    {
                        DestinationCidrBlock: '0.0.0.0/0',
                        GatewayId: 'igw-xxxxxx',
                        Origin: 'CreateRoute',
                        State: 'active',
                    },
                ],
            },
            {
                Associations: [{ SubnetId: 'sub-789012' }],
                RouteTableId: 'rtb-789012',
                Routes: [
                    {
                        DestinationCidrBlock: '1.1.2.1/24',
                        GatewayId: 'local',
                        Origin: 'CreateRouteTable',
                        State: 'active',
                    },
                    {
                        DestinationCidrBlock: '0.0.0.0/0',
                        NatGatewayId: 'nat-xxxxxx',
                        Origin: 'CreateRoute',
                        State: 'active',
                    },
                ],
            },
        ],
    })
        .on(client_ec2_1.DescribeVpnGatewaysCommand)
        .resolves({
        VpnGateways: [{ VpnGatewayId: 'gw-abcdef' }],
    });
});
test('looks up the requested VPC', async () => {
    // GIVEN
    const filter = { foo: 'bar' };
    const provider = new vpcs_1.VpcNetworkContextProviderPlugin(mockSDK);
    // WHEN
    const result = await provider.getValue({
        account: '123456789012',
        region: 'us-east-1',
        filter,
    });
    // THEN
    expect(result).toEqual({
        vpcId: 'vpc-1234567',
        vpcCidrBlock: '1.1.1.1/16',
        ownerAccountId: '123456789012',
        availabilityZones: ['bermuda-triangle-1337'],
        privateSubnetIds: ['sub-789012'],
        privateSubnetNames: ['Private'],
        privateSubnetRouteTableIds: ['rtb-789012'],
        publicSubnetIds: ['sub-123456'],
        publicSubnetNames: ['Public'],
        publicSubnetRouteTableIds: ['rtb-123456'],
        vpnGatewayId: 'gw-abcdef',
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpcsCommand, {
        Filters: [{ Name: 'foo', Values: ['bar'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeSubnetsCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeRouteTablesCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpnGatewaysCommand, {
        Filters: [
            { Name: 'attachment.vpc-id', Values: ['vpc-1234567'] },
            { Name: 'attachment.state', Values: ['attached'] },
            { Name: 'state', Values: ['available'] },
        ],
    });
});
test('throws when no such VPC is found', async () => {
    // GIVEN
    const filter = { foo: 'bar' };
    const provider = new vpcs_1.VpcNetworkContextProviderPlugin(mockSDK);
    mock_sdk_1.mockEC2Client.on(client_ec2_1.DescribeVpcsCommand).resolves({});
    // WHEN
    await expect(provider.getValue({
        account: '123456789012',
        region: 'us-east-1',
        filter,
    })).rejects.toThrow(/Could not find any VPCs matching/);
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpcsCommand, {
        Filters: [{ Name: 'foo', Values: ['bar'] }],
    });
});
test('throws when subnet with subnetGroupNameTag not found', async () => {
    // GIVEN
    const filter = { foo: 'bar' };
    const provider = new vpcs_1.VpcNetworkContextProviderPlugin(mockSDK);
    // WHEN
    await expect(provider.getValue({
        account: '123456789012',
        region: 'us-east-1',
        subnetGroupNameTag: 'DOES_NOT_EXIST',
        filter,
    })).rejects.toThrow(/Invalid subnetGroupNameTag: Subnet .* does not have an associated tag with Key='DOES_NOT_EXIST'/);
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpcsCommand, {
        Filters: [{ Name: 'foo', Values: ['bar'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeSubnetsCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeRouteTablesCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
});
test('does not throw when subnet with subnetGroupNameTag is found', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client.on(client_ec2_1.DescribeSubnetsCommand).resolves({
        Subnets: [
            {
                SubnetId: 'sub-123456',
                AvailabilityZone: 'bermuda-triangle-1337',
                MapPublicIpOnLaunch: true,
                Tags: [{ Key: 'DOES_EXIST', Value: 'SubnetName1' }],
            },
            {
                SubnetId: 'sub-789012',
                AvailabilityZone: 'bermuda-triangle-1337',
                MapPublicIpOnLaunch: false,
                Tags: [{ Key: 'DOES_EXIST', Value: 'SubnetName2' }],
            },
        ],
    });
    const filter = { foo: 'bar' };
    const provider = new vpcs_1.VpcNetworkContextProviderPlugin(mockSDK);
    // WHEN
    const result = await provider.getValue({
        account: '123456789012',
        region: 'us-east-1',
        subnetGroupNameTag: 'DOES_EXIST',
        filter,
    });
    // THEN
    expect(result).toEqual({
        vpcId: 'vpc-1234567',
        vpcCidrBlock: '1.1.1.1/16',
        ownerAccountId: '123456789012',
        availabilityZones: ['bermuda-triangle-1337'],
        privateSubnetIds: ['sub-789012'],
        privateSubnetNames: ['SubnetName2'],
        privateSubnetRouteTableIds: ['rtb-789012'],
        publicSubnetIds: ['sub-123456'],
        publicSubnetNames: ['SubnetName1'],
        publicSubnetRouteTableIds: ['rtb-123456'],
        vpnGatewayId: 'gw-abcdef',
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpcsCommand, {
        Filters: [{ Name: 'foo', Values: ['bar'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeSubnetsCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeRouteTablesCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpnGatewaysCommand, {
        Filters: [
            { Name: 'attachment.vpc-id', Values: ['vpc-1234567'] },
            { Name: 'attachment.state', Values: ['attached'] },
            { Name: 'state', Values: ['available'] },
        ],
    });
});
test('throws when multiple VPCs are found', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client.on(client_ec2_1.DescribeVpcsCommand).resolves({
        Vpcs: [{ VpcId: 'vpc-1' }, { VpcId: 'vpc-2' }],
    });
    const filter = { foo: 'bar' };
    const provider = new vpcs_1.VpcNetworkContextProviderPlugin(mockSDK);
    // WHEN
    await expect(provider.getValue({
        account: '123456789012',
        region: 'us-east-1',
        filter,
    })).rejects.toThrow(/Found 2 VPCs matching/);
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpcsCommand, {
        Filters: [{ Name: 'foo', Values: ['bar'] }],
    });
});
test('uses the VPC main route table when a subnet has no specific association', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client.on(client_ec2_1.DescribeRouteTablesCommand).resolves({
        RouteTables: [
            {
                Associations: [{ SubnetId: 'sub-123456' }],
                RouteTableId: 'rtb-123456',
                Routes: [
                    {
                        DestinationCidrBlock: '1.1.1.1/24',
                        GatewayId: 'local',
                        Origin: 'CreateRouteTable',
                        State: 'active',
                    },
                    {
                        DestinationCidrBlock: '0.0.0.0/0',
                        GatewayId: 'igw-xxxxxx',
                        Origin: 'CreateRoute',
                        State: 'active',
                    },
                ],
            },
            {
                Associations: [{ Main: true }],
                RouteTableId: 'rtb-789012',
                Routes: [
                    {
                        DestinationCidrBlock: '1.1.2.1/24',
                        GatewayId: 'local',
                        Origin: 'CreateRouteTable',
                        State: 'active',
                    },
                    {
                        DestinationCidrBlock: '0.0.0.0/0',
                        NatGatewayId: 'nat-xxxxxx',
                        Origin: 'CreateRoute',
                        State: 'active',
                    },
                ],
            },
        ],
    });
    const filter = { foo: 'bar' };
    const provider = new vpcs_1.VpcNetworkContextProviderPlugin(mockSDK);
    // WHEN
    const result = await provider.getValue({
        account: '123456789012',
        region: 'us-east-1',
        filter,
    });
    // THEN
    expect(result).toEqual({
        vpcId: 'vpc-1234567',
        vpcCidrBlock: '1.1.1.1/16',
        ownerAccountId: '123456789012',
        availabilityZones: ['bermuda-triangle-1337'],
        privateSubnetIds: ['sub-789012'],
        privateSubnetNames: ['Private'],
        privateSubnetRouteTableIds: ['rtb-789012'],
        publicSubnetIds: ['sub-123456'],
        publicSubnetNames: ['Public'],
        publicSubnetRouteTableIds: ['rtb-123456'],
        vpnGatewayId: 'gw-abcdef',
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpcsCommand, {
        Filters: [{ Name: 'foo', Values: ['bar'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeSubnetsCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeRouteTablesCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpnGatewaysCommand, {
        Filters: [
            { Name: 'attachment.vpc-id', Values: ['vpc-1234567'] },
            { Name: 'attachment.state', Values: ['attached'] },
            { Name: 'state', Values: ['available'] },
        ],
    });
});
test('Recognize public subnet by route table', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client
        .on(client_ec2_1.DescribeSubnetsCommand)
        .resolves({
        Subnets: [{ SubnetId: 'sub-123456', AvailabilityZone: 'bermuda-triangle-1337', MapPublicIpOnLaunch: false }],
    })
        .on(client_ec2_1.DescribeRouteTablesCommand)
        .resolves({
        RouteTables: [
            {
                Associations: [{ SubnetId: 'sub-123456' }],
                RouteTableId: 'rtb-123456',
                Routes: [
                    {
                        DestinationCidrBlock: '10.0.2.0/26',
                        Origin: 'CreateRoute',
                        State: 'active',
                        VpcPeeringConnectionId: 'pcx-xxxxxx',
                    },
                    {
                        DestinationCidrBlock: '10.0.1.0/24',
                        GatewayId: 'local',
                        Origin: 'CreateRouteTable',
                        State: 'active',
                    },
                    {
                        DestinationCidrBlock: '0.0.0.0/0',
                        GatewayId: 'igw-xxxxxx',
                        Origin: 'CreateRoute',
                        State: 'active',
                    },
                ],
            },
        ],
    })
        .on(client_ec2_1.DescribeVpnGatewaysCommand)
        .resolves({});
    const filter = { foo: 'bar' };
    const provider = new vpcs_1.VpcNetworkContextProviderPlugin(mockSDK);
    // WHEN
    const result = await provider.getValue({
        account: '123456789012',
        region: 'us-east-1',
        filter,
    });
    // THEN
    expect(result).toEqual({
        vpcId: 'vpc-1234567',
        vpcCidrBlock: '1.1.1.1/16',
        ownerAccountId: '123456789012',
        availabilityZones: ['bermuda-triangle-1337'],
        publicSubnetIds: ['sub-123456'],
        publicSubnetNames: ['Public'],
        publicSubnetRouteTableIds: ['rtb-123456'],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpcsCommand, {
        Filters: [{ Name: 'foo', Values: ['bar'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeSubnetsCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeRouteTablesCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpnGatewaysCommand, {
        Filters: [
            { Name: 'attachment.vpc-id', Values: ['vpc-1234567'] },
            { Name: 'attachment.state', Values: ['attached'] },
            { Name: 'state', Values: ['available'] },
        ],
    });
});
test('Recognize private subnet by route table with NAT Gateway', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client
        .on(client_ec2_1.DescribeSubnetsCommand)
        .resolves({
        Subnets: [{ SubnetId: 'sub-123456', AvailabilityZone: 'bermuda-triangle-1337', MapPublicIpOnLaunch: false }],
    })
        .on(client_ec2_1.DescribeRouteTablesCommand)
        .resolves({
        RouteTables: [
            {
                Associations: [{ SubnetId: 'sub-123456' }],
                RouteTableId: 'rtb-123456',
                Routes: [
                    {
                        DestinationCidrBlock: '10.0.2.0/26',
                        Origin: 'CreateRoute',
                        State: 'active',
                        VpcPeeringConnectionId: 'pcx-xxxxxx',
                    },
                    {
                        DestinationCidrBlock: '10.0.1.0/24',
                        GatewayId: 'local',
                        Origin: 'CreateRouteTable',
                        State: 'active',
                    },
                    {
                        DestinationCidrBlock: '0.0.0.0/0',
                        NatGatewayId: 'nat-xxxxxx',
                        Origin: 'CreateRoute',
                        State: 'active',
                    },
                ],
            },
        ],
    })
        .on(client_ec2_1.DescribeVpnGatewaysCommand)
        .resolves({});
    const filter = { foo: 'bar' };
    const provider = new vpcs_1.VpcNetworkContextProviderPlugin(mockSDK);
    // WHEN
    const result = await provider.getValue({
        account: '123456789012',
        region: 'us-east-1',
        filter,
    });
    // THEN
    expect(result).toEqual({
        vpcId: 'vpc-1234567',
        vpcCidrBlock: '1.1.1.1/16',
        ownerAccountId: '123456789012',
        availabilityZones: ['bermuda-triangle-1337'],
        privateSubnetIds: ['sub-123456'],
        privateSubnetNames: ['Private'],
        privateSubnetRouteTableIds: ['rtb-123456'],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpcsCommand, {
        Filters: [{ Name: 'foo', Values: ['bar'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeSubnetsCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeRouteTablesCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpnGatewaysCommand, {
        Filters: [
            { Name: 'attachment.vpc-id', Values: ['vpc-1234567'] },
            { Name: 'attachment.state', Values: ['attached'] },
            { Name: 'state', Values: ['available'] },
        ],
    });
});
test('Recognize private subnet by route table with Transit Gateway', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client
        .on(client_ec2_1.DescribeSubnetsCommand)
        .resolves({
        Subnets: [{ SubnetId: 'sub-123456', AvailabilityZone: 'bermuda-triangle-1337', MapPublicIpOnLaunch: false }],
    })
        .on(client_ec2_1.DescribeRouteTablesCommand)
        .resolves({
        RouteTables: [
            {
                Associations: [{ SubnetId: 'sub-123456' }],
                RouteTableId: 'rtb-123456',
                Routes: [
                    {
                        DestinationCidrBlock: '10.0.2.0/26',
                        Origin: 'CreateRoute',
                        State: 'active',
                        VpcPeeringConnectionId: 'pcx-xxxxxx',
                    },
                    {
                        DestinationCidrBlock: '10.0.1.0/24',
                        GatewayId: 'local',
                        Origin: 'CreateRouteTable',
                        State: 'active',
                    },
                    {
                        DestinationCidrBlock: '0.0.0.0/0',
                        TransitGatewayId: 'tgw-xxxxxx',
                        Origin: 'CreateRoute',
                        State: 'active',
                    },
                ],
            },
        ],
    })
        .on(client_ec2_1.DescribeVpnGatewaysCommand)
        .resolves({});
    const filter = { foo: 'bar' };
    const provider = new vpcs_1.VpcNetworkContextProviderPlugin(mockSDK);
    // WHEN
    const result = await provider.getValue({
        account: '123456789012',
        region: 'us-east-1',
        filter,
    });
    // THEN
    expect(result).toEqual({
        vpcId: 'vpc-1234567',
        vpcCidrBlock: '1.1.1.1/16',
        ownerAccountId: '123456789012',
        availabilityZones: ['bermuda-triangle-1337'],
        privateSubnetIds: ['sub-123456'],
        privateSubnetNames: ['Private'],
        privateSubnetRouteTableIds: ['rtb-123456'],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpcsCommand, {
        Filters: [{ Name: 'foo', Values: ['bar'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeSubnetsCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeRouteTablesCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpnGatewaysCommand, {
        Filters: [
            { Name: 'attachment.vpc-id', Values: ['vpc-1234567'] },
            { Name: 'attachment.state', Values: ['attached'] },
            { Name: 'state', Values: ['available'] },
        ],
    });
});
test('Recognize isolated subnet by route table', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client
        .on(client_ec2_1.DescribeSubnetsCommand)
        .resolves({
        Subnets: [{ SubnetId: 'sub-123456', AvailabilityZone: 'bermuda-triangle-1337', MapPublicIpOnLaunch: false }],
    })
        .on(client_ec2_1.DescribeRouteTablesCommand)
        .resolves({
        RouteTables: [
            {
                Associations: [{ SubnetId: 'sub-123456' }],
                RouteTableId: 'rtb-123456',
                Routes: [
                    {
                        DestinationCidrBlock: '10.0.1.0/24',
                        GatewayId: 'local',
                        Origin: 'CreateRouteTable',
                        State: 'active',
                    },
                ],
            },
        ],
    })
        .on(client_ec2_1.DescribeVpnGatewaysCommand)
        .resolves({});
    const filter = { foo: 'bar' };
    const provider = new vpcs_1.VpcNetworkContextProviderPlugin(mockSDK);
    // WHEN
    const result = await provider.getValue({
        account: '123456789012',
        region: 'us-east-1',
        filter,
    });
    // THEN
    expect(result).toEqual({
        vpcId: 'vpc-1234567',
        vpcCidrBlock: '1.1.1.1/16',
        ownerAccountId: '123456789012',
        availabilityZones: ['bermuda-triangle-1337'],
        isolatedSubnetIds: ['sub-123456'],
        isolatedSubnetNames: ['Isolated'],
        isolatedSubnetRouteTableIds: ['rtb-123456'],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpcsCommand, {
        Filters: [{ Name: 'foo', Values: ['bar'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeSubnetsCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeRouteTablesCommand, {
        Filters: [{ Name: 'vpc-id', Values: ['vpc-1234567'] }],
    });
    expect(mock_sdk_1.mockEC2Client).toHaveReceivedCommandWith(client_ec2_1.DescribeVpnGatewaysCommand, {
        Filters: [
            { Name: 'attachment.vpc-id', Values: ['vpc-1234567'] },
            { Name: 'attachment.state', Values: ['attached'] },
            { Name: 'state', Values: ['available'] },
        ],
    });
});
